Automattic/newspack-plugin
# feat: add support for reCAPTCHA v2

### All Submissions:






### Changes proposed in this Pull Request:



Adds optional support for reCAPTCHA v2 ([invisible](https://developers.google.com/recaptcha/docs/invisible) flavor only) in addition to the already-supported v3. 

<img width="1059" alt="Screenshot 2024-05-23 at 2 42 23 PM" src="https://github.com/Automattic/newspack-plugin/assets/2230142/2ed40e74-3781-48f4-b325-5c93c05d0891">

@Automattic/newspack-design I could use a design review from you on the new admin UI for managing the different versions, as well as the positioning of the checkbox widget in the different contexts (screenshots below).

Closes `1206943664367857/1207301960751653`.

### How to test the changes in this Pull Request:

1. Start with a site that already has reCAPTCHA v3 configured and working. Check out this branch as well as https://github.com/Automattic/newspack-newsletters/pull/1514 and https://github.com/Automattic/newspack-blocks/pull/1744.

2. Visit [Google's reCAPTCHA admin](https://www.google.com/recaptcha/admin/create) and create a new account: 
  - Type set to **Challenge (v2)** + **Invisible reCAPTCHA badge**
  - After creating, go to settings and add your test site's URL as a domain. You can also optionally bump the "Security" slider all the way to the right to make it more likely that you'll see the reCAPTCHA challenge on the front-end.

<img width="731" alt="Screenshot 2024-05-23 at 2 44 50 PM" src="https://github.com/Automattic/newspack-plugin/assets/2230142/93090d1d-c866-4be8-bf56-7d96aaec0438">

3. In WP admin, visit **Newspack > Connections** and scroll down to the reCAPTCHA section. Confirm the new "reCAPTCHA version" option, and that it's set to v3 since you had it configured already. Leave v3 enabled for now.

4. As a reader, perform the following actions, each in a new session, and confirm that you're able to complete each one:
  - Register via the Registration block
  - Register via the "Sign In" auth modal
  - Log into an existing account via the "Sign In" auth modal (note: login will not be checked by reCAPTCHA)
  - Sign up for newsletter lists via the Newsletter Signup block
  - Purchase a product via the modal checkout (Donate or Checkout Button blocks)
  - **Note:** For v2 invisible and v3 reCAPTCHA types, you may not be presented with any challenge or feedback if the initial check succeeds. To confirm that you're being tested and verified by reCAPTCHA, you can add `error_log( print_r( $captcha_verify, true ) );` just after [this line](https://github.com/Automattic/newspack-plugin/blob/feat/recaptcha-v2/includes/class-recaptcha.php#L364) to see the results of the verification request.

5. Back in WP admin, select the "v2 invisible" option (confirming that the v3-only "Threshold" option disappears) and paste your site key + secret from the invisible account you created in step 2. Save.

6. Repeat the actions from step 4 as a reader and confirm you're able to complete each one.
  - For v2 invisible reCAPTCHA, you may be asked to complete a challenge like this if the initial verification fails upon form submission. Once you complete the challenge successfully, the form should automatically submit.

<img width="853" alt="Screenshot 2024-05-22 at 12 21 48 PM" src="https://github.com/Automattic/newspack-plugin/assets/2230142/1e7c58b1-e1ae-4b28-9456-0f6da4fd4cbf">

7. Smoke test: disable reCAPTCHA in the Connections wizard by toggling it off, and confirm you can still complete all actions as a reader.

### Other information:




---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1207301960751653
